api/sendfild/service/service.go

```diff
@@ -0,0 +1,713 @@
+package service
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math"
+	"math/big"
+	"net"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/lotus/chain/types"
+	"github.com/ipfs/go-cid"
+	logging "github.com/ipfs/go-log/v2"
+	"github.com/textileio/go-threads/util"
+	"github.com/textileio/powergate/v2/lotus"
+	pb "github.com/textileio/textile/v2/api/sendfild/pb"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+const (
+	collectionName = "sendfil"
+)
+
+var _ pb.SendFilServiceServer = (*Service)(nil)
+
+var log = logging.Logger("sendfil")
+
+type msgCid struct {
+	Cid       string    `bson:"cid"`
+	CreatedAt time.Time `bson:"created_at"`
+}
+
+type txn struct {
+	ID            primitive.ObjectID `bson:"_id"`
+	From          string             `bson:"from"`
+	To            string             `bson:"to"`
+	AmountNanoFil int64              `bson:"amount_nano_fil"`
+	MessageCids   []msgCid           `bson:"message_cids"`
+	MessageState  pb.MessageState    `bson:"message_state"`
+	Waiting       bool               `bson:"waiting"`
+	FailureMsg    string             `bson:"failure_msg"`
+	CreatedAt     time.Time          `bson:"created_at"`
+	UpdatedAt     time.Time          `bson:"updated_at"`
+}
+
+func (t txn) latestMsgCid() (msgCid, error) {
+	if len(t.MessageCids) == 0 {
+		return msgCid{}, fmt.Errorf("no message cids found")
+	}
+	return t.MessageCids[len(t.MessageCids)-1], nil
+}
+
+type Service struct {
+	clientBuilder lotus.ClientBuilder
+	col           *mongo.Collection
+	server        *grpc.Server
+	waiting       map[primitive.ObjectID]chan waitResult
```

Comment:
We keep an in memory map of txn mongo ids to channels that complete once lotus tells us the message is on chain (or the message fails or we timeout the waiting)

Reply:
This lets us share any message waiting between multiple calls and doesn't create additional calls to lotus to do that.